OSDOCS-2396: Add hardware_enablement category and SRO documentation (#36674)
openshift/openshift-docs: modules/psap-special-resource-operator-using-configmaps.adoc

+
[source,terminal]
----
$ oc get pods -n simple-kmod
----
+
.Example output
[source,terminal]
----
NAME                                                  READY   STATUS      RESTARTS   AGE
simple-kmod-driver-build-12813789169ac0ee-1-build     0/1     Completed   0          7m12s
simple-kmod-driver-container-12813789169ac0ee-mjsnh   1/1     Running     0          8m2s
simple-kmod-driver-container-12813789169ac0ee-qtkff   1/1     Running     0          8m2s
----

. Use the `oc logs` command, along with the build pod name obtained from the `oc get pods` command above, to display the logs of the simple-kmod driver container image build:
+
[source,terminal]
----
$ oc logs pod/simple-kmod-driver-build-12813789169ac0ee-1-build -n simple-kmod
----

. To verify that the simple-kmod kernel modules are loaded, execute the `lsmod` command in one of the driver container pods that was returned from the `oc get pods` command above. 
+
[source,terminal]
----
$ oc exec -n simple-kmod -it pod/simple-kmod-driver-container-12813789169ac0ee-mjsnh -- lsmod | grep simple
----
+
.Example output
[source,terminal]
----
simple_procfs_kmod     16384  0
simple_kmod            16384  0
----

> [NOTE]

Reviewer: Add an enter between line 305 and 306 so the note displays properly. If you want to keep it together with the Example output, then add a + before the NOTE.

// Module included in the following assemblies:
//
// * hardware_enablement/psap-special-resource-operator.adoc

[id="deploy-simple-kmod-using-configmap-chart"]
= Building and running the simple-kmod SpecialResource by using a ConfigMap

> In this example, the simple-kmod kernel module is used to show how the loading of a kernel module can be managed by SRO, by storing Helm chart templates in a ConfigMap.

Reviewer 2: I think this could be worded in a less awkward / passive way.

```suggestion
In this example, the simple-kmod kernel module is used to show how SRO can manage a driver container which is defined in Helm chart templates stored in a ConfigMap.
```

Author: fixed!

.Prerequisites

* You have a running {product-title} cluster.
* You set the Image Registry Operator state to Managed for your cluster.
* You installed the OpenShift CLI (`oc`).
* You are logged into the OpenShift CLI as a user with `cluster-admin` privileges.
* You installed Node Feature Discovery (NFD) Operator.
* You installed the Special Resource Operator.
* You installed the Helm CLI (`helm`).

.Procedure
. To create a simple-kmod `SpecialResource` object, define an ImageStream and BuildConfig to build the image, and a ServiceAccount, Role, RoleBinding, and DaemonSet to run the container. The ServiceAccount, Role, and RoleBinding are required to run the DaemonSet with the privileged securityContext so that the kernel module can be loaded.
.. Create a `templates` directory, and change into it:
+
[source,terminal]
----
$ mkdir -p chart/simple-kmod-0.0.1/templates
----
+
[source,terminal]
----
$ cd chart/simple-kmod-0.0.1/templates
----

> .. Save this YAML template for the ImageStream and BuildConfig in the `templates` directory as `0000-buildconfig.yaml`. Note that the templates such as `{{.Values.specialresource.metadata.name}}` are filled in by SRO, based on fields in the `SpecialResource` CR and variables known to the Operator such as `{{.Values.KernelFullVersion}}`

Reviewer 3: I'd consider adding your "Note that" information in a callout that points to the these lines int he YAML below. You can always have a single callout (`<1>`) below the code block, and use it more than once on each line that is applicable. 

Also, just to clarify, because the SRO is filling them in, they should leave these lines alone and not change them at all, correct?

              command: ["/bin/sh", "-c", "systemctl stop kmods-via-containers@{{.Values.specialresource.metadata.name}}"]
        securityContext:
          privileged: true
      nodeSelector:
        node-role.kubernetes.io/worker: ""
        feature.node.kubernetes.io/kernel-version.full: "{{.Values.KernelFullVersion}}"
----

.. Change into the `chart/simple-kmod-0.0.1` directory:
+
[source, terminal]
----
$ cd ..
----

.. Save the following YAML for the Chart as `Chart.yaml` in the `chart/simple-kmod-0.0.1` directory:
+
[source, yaml]
----
apiVersion: v2
name: simple-kmod
description: Simple kmod will deploy a simple kmod driver-container
icon: https://avatars.githubusercontent.com/u/55542927
type: application
version: 0.0.1
appVersion: 1.0.0
----

. From the `chart` directory, create the chart using the `helm package` command:
+
[source, terminal]
----
$ helm package simple-kmod-0.0.1/
----
+
.Example output
[source,terminal]
----
Successfully packaged chart and saved it to: /data/dagray/git/dagrayvid/special-resource-operator/yaml-for-docs/chart/simple-kmod-0.0.1/simple-kmod-0.0.1.tgz
----

. Create a configuration map to store the chart files:
+
[source,terminal]
----

> $ mkdir cm

Reviewer 2: Sure. Here are the substep descriptions and steps:

Create a directory for the files to be stored in the ConfigMap
`$ mkdir cm`

Copy the Helm chart into the `cm` directory.
`$ cp simple-kmod-0.0.1.tgz cm/simple-kmod-0.0.1.tgz`

Create an index file specifying the Helm repo which contains the Helm chart:
`$ helm repo index cm --url=cm://simple-kmod/simple-kmod-chart`

Create a namespace for the objects defined in the Helm chart:
`$ oc create namespace simple-kmod`

Create the ConfigMap object:
` $ oc create cm simple-kmod-chart --from-file=cm/index.yaml --from-file=cm/simple-kmod-0.0.1.tgz -n simple-kmod`

----
$ oc create cm simple-kmod-chart --from-file=cm/index.yaml --from-file=cm/simple-kmod-0.0.1.tgz -n simple-kmod
----

. Use the following `SpecialResource` manifest to deploy the simple-kmod object using the Helm chart that you created in the ConfigMap. Save this YAML as `simple-kmod-configmap.yaml`:
+
[source,yaml]
----
apiVersion: sro.openshift.io/v1beta1
kind: SpecialResource
metadata:
  name: simple-kmod
spec:
  #debug: true <1>
  namespace: simple-kmod
  chart:
    name: simple-kmod
    version: 0.0.1
    repository:
      name: example
      url: cm://simple-kmod/simple-kmod-chart <2>
  set:
    kind: Values
    apiVersion: sro.openshift.io/v1beta1
    kmodNames: ["simple-kmod", "simple-procfs-kmod"]
    buildArgs:
    - name: "KMODVER"
      value: "SRO"
  driverContainer:
    source:
      git:
        ref: "master"
        uri: "https://github.com/openshift-psap/kvc-simple-kmod.git"
----

<1> Optional: Uncomment the "#debug: true" line, to have the YAML files in the chart printed in full in the Operator logs and to verify that the logs are created and templated properly.
<2> The `spec.chart.repository.url` field tells SRO to look for the chart in a ConfigMap.

> . From a command line, create the `SpecialResource` file:

Author: Fixed - ty!

+
[source,terminal]
----
$ oc create -f simple-kmod-configmap.yaml
----

. The `simple-kmod` resources are deployed in the `simple-kmod` namespace as specified in the object manifest. After a short time, the build pod for the `simple-kmod` driver container starts running. The build completes after a few minutes, and then the driver container pods start running. 
+
Use `oc get pods` command to display the status of the build pods:
+
[source,terminal]
----
$ oc get pods -n simple-kmod
----
+
.Example output
[source,terminal]
----
NAME                                                  READY   STATUS      RESTARTS   AGE
simple-kmod-driver-build-12813789169ac0ee-1-build     0/1     Completed   0          7m12s
simple-kmod-driver-container-12813789169ac0ee-mjsnh   1/1     Running     0          8m2s
simple-kmod-driver-container-12813789169ac0ee-qtkff   1/1     Running     0          8m2s
----

. Use the `oc logs` command, along with the build pod name obtained from the `oc get pods` command above, to display the logs of the simple-kmod driver container image build:
+
[source,terminal]
----
$ oc logs pod/simple-kmod-driver-build-12813789169ac0ee-1-build -n simple-kmod
----

> . To verify that the simple-kmod kernel modules are loaded, execute the `lsmod` command in one of the driver container pods that was returned from the `oc get pods` command above. 

Reviewer 3: Period should be a colon to be consistent with the other steps

  name: {{.Values.specialresource.metadata.name}}-{{.Values.groupName.driverBuild}} <1>
  annotations:
    specialresource.openshift.io/wait: "true"
    specialresource.openshift.io/driver-container-vendor: simple-kmod
    specialresource.openshift.io/kernel-affine: "true"
spec:
  nodeSelector:
    node-role.kubernetes.io/worker: ""
  runPolicy: "Serial"
  triggers:
    - type: "ConfigChange"
    - type: "ImageChange"
  source:
    git:
      ref: {{.Values.specialresource.spec.driverContainer.source.git.ref}}
      uri: {{.Values.specialresource.spec.driverContainer.source.git.uri}}
    type: Git
  strategy:
    dockerStrategy:
      dockerfilePath: Dockerfile.SRO
      buildArgs:
        - name: "IMAGE"
          value: {{ .Values.driverToolkitImage  }}
        {{- range $arg := .Values.buildArgs }}
        - name: {{ $arg.name }}
          value: {{ $arg.value }}
        {{- end }}
        - name: KVER
          value: {{ .Values.kernelFullVersion }}
  output:
    to:
      kind: ImageStreamTag
      name: {{.Values.specialresource.metadata.name}}-{{.Values.groupName.driverContainer}}:v{{.Values.kernelFullVersion}} <1>
----

> <1> The templates such as `{{.Values.specialresource.metadata.name}}` are filled in by SRO, based on fields in the `SpecialResource` CR and variables known to the Operator such as `{{.Values.KernelFullVersion}}`

Reviewer 3: Missing a period

subjects:
- kind: ServiceAccount
  name: {{.Values.specialresource.metadata.name}}-{{.Values.groupName.driverContainer}}
  namespace: {{.Values.specialresource.spec.namespace}}
---
apiVersion: apps/v1
kind: DaemonSet
metadata:
  labels:
    app: {{.Values.specialresource.metadata.name}}-{{.Values.groupName.driverContainer}}
  name: {{.Values.specialresource.metadata.name}}-{{.Values.groupName.driverContainer}}
  annotations:
    specialresource.openshift.io/wait: "true"
    specialresource.openshift.io/state: "driver-container"
    specialresource.openshift.io/driver-container-vendor: simple-kmod
    specialresource.openshift.io/kernel-affine: "true"
spec:
  updateStrategy:
    type: OnDelete
  selector:
    matchLabels:
      app: {{.Values.specialresource.metadata.name}}-{{.Values.groupName.driverContainer}}
  template:
    metadata:
      # Mark this pod as a critical add-on; when enabled, the critical add-on scheduler
      # reserves resources for critical add-on pods so that they can be rescheduled after
      # a failure.  This annotation works in tandem with the toleration below.
      annotations:

>         scheduler.alpha.kubernetes.io/critical-pod: ""

Reviewer 4: The annotation is non-functional starting kubernetes 1.16 (going way before OCP 4.6)

        lifecycle:
          preStop:
            exec:
              command: ["/bin/sh", "-c", "systemctl stop kmods-via-containers@{{.Values.specialresource.metadata.name}}"]
        securityContext:
          privileged: true
      nodeSelector:
        node-role.kubernetes.io/worker: ""
        feature.node.kubernetes.io/kernel-version.full: "{{.Values.KernelFullVersion}}"
----

.. Change into the `chart/simple-kmod-0.0.1` directory:
+
[source, terminal]
----
$ cd ..
----

.. Save the following YAML for the chart as `Chart.yaml` in the `chart/simple-kmod-0.0.1` directory:
+
[source, yaml]
----
apiVersion: v2
name: simple-kmod
description: Simple kmod will deploy a simple kmod driver-container
icon: https://avatars.githubusercontent.com/u/55542927
type: application
version: 0.0.1
appVersion: 1.0.0
----

. From the `chart` directory, create the chart using the `helm package` command:
+
[source, terminal]
----
$ helm package simple-kmod-0.0.1/
----
+
.Example output
[source,terminal]
----
Successfully packaged chart and saved it to: /data/<username>/git/<github_username>/special-resource-operator/yaml-for-docs/chart/simple-kmod-0.0.1/simple-kmod-0.0.1.tgz
----

> . Create a configuration map to store the chart files:

Reviewer 3: This wasn't resolved:
configuration map -> config map
